### PR TITLE
Moved flatMap intro SignalProducerType extension.

### PR DESCRIPTION
**Pull request by nikita-leonov**
According to @neilpa message in #2377 there was language and compiler limitations that didn't allowed this before. On a Xcode 7.0 (7A220) it complies well and all unit-tests are passing. Does it mean it can me merged in, or there is a special case that still may crash such implementation?

### CODE REVIEW

File: ReactiveCocoa/Swift/SignalProducer.swift [hunk 1]
```diff
@@ -913,7 +913,35 @@ public func zip<S: SequenceType, T, Error where S.Generator.Element == SignalPro
...
+	/// Flattens the inner producers sent upon `producer` (into a single producer of
+	/// values), according to the semantics of the given strategy.
+	///
+	/// If `producer` or an active inner producer emits an error, the returned
+	/// producer will forward that error immediately.
+	///
+	/// `Interrupted` events on inner producers will be treated like `Completed`
+	/// events on inner producers.
+	@warn_unused_result(message="Did you forget to call `start` on the producer?")
+	public func flatten(strategy: FlattenStrategy) -> SignalProducer<T.T, E> {
```
Comment (NachoSoto) [inline ReactiveCocoa/Swift/SignalProducer.swift]: Can this be moved to `SignalProducerType` too?

Comment (nikita-leonov) [inline ReactiveCocoa/Swift/SignalProducer.swift]: @NachoSoto good question, let me check, as I was trying to solve only my case and didn't thought that someone may require flatten as well.

Comment (neilpa) [inline ReactiveCocoa/Swift/SignalProducer.swift]: I'll be pleasantly surprised if it works w/out making the compiler die. IIRC I tried to convert this in multiple betas up to 5 and was disappointed each time.

Comment (neilpa) [inline ReactiveCocoa/Swift/SignalProducer.swift]: If it does work, you should also be able to change `filterNil` because that was the other problem operator.

Comment (nikita-leonov) [inline ReactiveCocoa/Swift/SignalProducer.swift]: @neilpa what sort of issue did you have with compiler? In my case I just have an issues with types in ```lift``` call which looks like a legit call:
```
Cannot invoke 'lift' with an argument list of type '((Signal<SignalProducer<Self.T, Self.E>, Self.E>) -> Signal<Self.T, Self.E>)
```

Comment (neilpa) [inline ReactiveCocoa/Swift/SignalProducer.swift]: Previously it would seg-fault the compiler. That looks like a normal compiler error to me.

Comment (neilpa) [inline ReactiveCocoa/Swift/SignalProducer.swift]: I recommend storing intermediate steps in locals with explicit type annotations. That's usually how I track down the actual source of errors when there's otherwise a lot of type-inference involved.

Comment (nikita-leonov) [inline ReactiveCocoa/Swift/SignalProducer.swift]: Hm, yes I was over-engineering. Actual change was as simple as replace ```SignalProducer``` to ```SignalProducerType``` and retaining the rest for flatten as is. it compiles, it works as it pass the tests.

There are many other things that could be moved to ```SignalProducerType``` and ```SignalType``` in the similar manner, but I would prefer to keep it in a separate PR in other case it will grow to big.

**Comment by nikita-leonov** [comment]
Actually wait a little bit and I move reminders in ```SignalProducerType``` and ```SignalType```. It seems it is not as big as I expected.

File: ReactiveCocoa/Swift/SignalProducer.swift (hunk 1 above)
Comment (nikita-leonov) [inline ReactiveCocoa/Swift/SignalProducer.swift]: ps practice of breaking down into smaller steps with an expected type clearly defined is a good one, I use it often as well. The problem in my case that there was no further way to breakdown... I actually found the place of an error but because of a very complex signature of signals etc was too blind to notice it.

**Comment by nikita-leonov** [comment]
I think it is done. By done I mean done-done :)

**Comment by neilpa** [comment]
Awesome! Everything's implemented in terms of protocol extensions now :tada:

**Comment by NachoSoto** [comment]
Awesome! :grinning:
